rust/psibase_macros/src/graphql_macro.rs

```diff
@@ -0,0 +1,241 @@
+use darling::ToTokens;
+use proc_macro::TokenStream;
+use proc_macro2::Ident;
+use proc_macro_error::abort;
+use quote::quote;
+use syn::{
+    parse::{Parse, ParseStream},
+    parse_macro_input, AttrStyle, ImplItem, Item, Result, Token, TypeParam, TypeParamBound,
+};
+
+#[derive(Debug)]
+struct Args {
+    query_name: Ident,
+    table: Ident,
+    key_fn: TypeParam,
+    subindex_params: Vec<TypeParam>,
+    subindex_remaining_key_ty: Option<TypeParamBound>,
+}
+
+impl Parse for Args {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let query_name = input.parse().expect("Query name not present");
+        input.parse::<Token![,]>().expect("expected ,");
+
+        let table = input.parse().expect("Table name not present");
+        input.parse::<Token![,]>().expect("expected ,");
+
+        let key_fn = input.parse().expect("Table key function not present");
+
+        let mut subindex_params = Vec::new();
+        let mut subindex_remaining_key_ty: Option<TypeParamBound> = None;
+
+        if input.parse::<Token![,]>().is_ok() {
+            loop {
+                if input.is_empty() {
+                    break;
+
+                // Every param has a colon to define its type
+                } else if input.peek2(Token![:]) {
+                    let param = input
+                        .parse::<TypeParam>()
+                        .expect("unable to parse subindex param type");
+                    subindex_params.push(param);
+                    if input.is_empty() {
+                        break;
+                    }
+                    input.parse::<Token![,]>().expect("expected ,");
+
+                // This is the last param, it should always be the remaining key TypeParamBound
```

Comment:
If this needs to be the last param, then it needs to check that it is at the end after an optional trailing comma.

Reply:
Fixed.